packages/arb-avm-cpp/data_storage/src/arbcore.cpp

```diff
@@ -1410,15 +1426,34 @@ void ArbCore::printCoreThreadBacktrace() {
 
 bool ArbCore::threadBody(ThreadDataStruct& thread_data) {
     bool isMachineValid;
-    {
+    if (core_machine->status() == MachineThread::MACHINE_ERROR) {
+        isMachineValid = false;
+        core_machine->clearError();
+    } else {
         ReadTransaction tx(data_storage);
         isMachineValid = isValid(tx, core_machine->getReorgData());
     }
     if (!isMachineValid) {
-        std::cerr << "Core thread operating on invalid machine. Rolling back."
-                  << std::endl;
-        assert(false);
-        auto status = reorgToMessageCountOrBefore(0, false, thread_data.cache);
+        if (coreConfig.thread_max_failure_count != 0 &&
+            thread_data.thread_failure_count >
+                coreConfig.thread_max_failure_count) {
+            std::cerr << "Core thread had too many errors, aborting"
+                      << std::endl;
+            return false;
+        }
+        rocksdb::Status status;
+        if (thread_data.thread_failure_count == 0) {
+            std::cerr << "Core thread operating on invalid machine, "
+                         "loading last saved checkpoint"
+                      << std::endl;
+            status = reorgToLastCheckpoint(thread_data.cache);
+        } else {
+            std::cerr << "Core thread operating on invalid machine, "
+                         "loading penultimate saved checkpoint, failure_count: "
+                      << thread_data.thread_failure_count << std::endl;
+            status = reorgToPenultimateCheckpoint(thread_data.cache);
```

Comment:
This might be overkill, but you could have a reorgToNthCheckpoint method which would also remove the branching here

Reply:
The it would have the potential of skipping more than a single checkpoint at a time.